vochain/state/sik.go

```diff
@@ -0,0 +1,274 @@
+package state
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/tree/arbo"
+)
+
+// TODO: Move the definition to the right place
+const SIKROOT_HYSTERESIS_BLOCKS = 32
+
+const (
+	// encodedHeightLen constant is the number of bytes of the encoded
+	// hysteresis that contains the hysteresis height value, starting from the
+	// last byte
+	encodedHeightLen = 4
+	// sikLeafValueLen constant contains the number of bytes that a leaf value
+	// has
+	sikLeafValueLen = 32
+)
+
+// SIK type abstracts a slice of bytes that contains the Secret Identity Key
+// value of a user
+type SIK []byte
+
+// SIKFromAddress function return the current SIK value associated to the provided
+// address.
+func (v *State) SIKFromAddress(address common.Address) (SIK, error) {
+	return v.mainTreeViewer(false).DeepGet(address.Bytes(), StateTreeCfg(TreeSIK))
+}
+
+// SetAddressSIK function creates or update the SIK of the provided address in the
+// state. It covers the following cases:
+//   - It checks if already exists a valid SIK for the provided address and if
+//     so it returns an error.
+//   - If no SIK exists for the provided address, it will create one with the
+//     value provided.
+//   - If it exists but it is not valid, overwrite the stored value with the
+//     provided one.
+func (v *State) SetAddressSIK(address common.Address, newSik SIK) error {
+	siksTree, err := v.Tx.DeepSubTree(StateTreeCfg(TreeSIK))
+	if err != nil {
+		return err
+	}
+	// check if exists a registered sik for the provided address, query also for
+	// no commited tree version
+	v.Tx.Lock()
+	rawSik, err := siksTree.Get(address.Bytes())
+	v.Tx.Unlock()
+	if errors.Is(err, arbo.ErrKeyNotFound) {
+		// if not exists create it
+		log.Debugw("setSIK (create)",
+			"address", address.String(),
+			"sik", SIK(newSik).String())
```

Comment:
If I am not wrong `newSik` is already a SIK type (type of the argument function)